pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +133,487 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+
+		res, err := r.processMaintModeOn(statusAnnoData, node)
+		return res, err
 
-type ReconcileHandler interface {
-	StartPodInformer(node *corev1.Node, stop <-chan struct{}) error
+	} else if statusAnnoData != nil { // spec deleted but annotation exists -> can be request to exit maintenance mode.
+
+		res, err := r.processMaintModeOff(statusAnnoData, node)
+		return res, err
+	}
+	return reconcile.Result{}, nil
 }
 
-// ReconcileNodeMaintenance reconciles a NodeMaintenance object
-type ReconcileNodeMaintenance struct {
-	// This client, initialized using mgr.Client() above, is a split client
-	// that reads objects from the cache and writes to the apiserver
-	client      client.Client
-	scheme      *runtime.Scheme
-	drainer     *drain.Helper
-	podInformer cache.SharedInformer
+func (r *ReconcileNodeMaintenance) adjustLeaseTimeIfRequired(specAnnoData *ReconcileNodeMaintenanceSpecInfo, statusAnnoData *ReconcileNodeMaintenanceStatusInfo, lease *coordv1beta1.Lease) error {
+	duration := statusAnnoData.OpDeadline.Sub(statusAnnoData.OpStart)
+	durationSeconds := int64(duration.Nanoseconds()) / int64(time.Second)
+
+	diffTime := int64(specAnnoData.RequestedMaintenanceWindowSeconds) - durationSeconds
+
+	if diffTime < -3 || diffTime > 3 {
+		r.reqLogger.Infof("Lease time has been changed. %d", int(durationSeconds))
+
+		specSeconds := int32(durationSeconds)
+		lease.Spec.LeaseDurationSeconds = &specSeconds
+
+		if err := r.client.Update(context.TODO(), lease); err != nil {
+
+			return err
+		}
+	}
+	return nil
 }
 
-var Handler ReconcileHandler
+func (r *ReconcileNodeMaintenance) processMaintModeOn(statusAnnoData *ReconcileNodeMaintenanceStatusInfo, node *corev1.Node) (reconcile.Result, error) {
 
-// Reconcile reads that state of the cluster for a NodeMaintenance object and makes changes based on the state read
-// and what is in the NodeMaintenance.Spec
-// Note:
-// The Controller will requeue the Request to be processed again if the returned error is non-nil or
-// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	reqLogger := log.WithFields(log.Fields{"Request.Namespace": request.Namespace, "Request.Name": request.Name})
-	reqLogger.Info("Reconciling NodeMaintenance")
+	nodeName := node.ObjectMeta.Name
 
-	// Fetch the NodeMaintenance instance
-	instance := &kubevirtv1alpha1.NodeMaintenance{}
-	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			// Request object not found, could have been deleted after reconcile request.
-			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
-			// Return and don't requeue
-			reqLogger.Infof("NodeMaintenance Object: %s Deleted ", request.NamespacedName)
-			return reconcile.Result{}, nil
+	switch statusAnnoData.Status {
+	case NodeStateInit:
```

Comment:
This state seems unnecessary

Reply:
You need it if the lease is initially owned by some other holder, in this case the request is requeued (to distinguish the case from a real error) (line #196 right now)